inc/Engine/Media/Fonts/Controller/Filesystem.php

```diff
@@ -0,0 +1,242 @@
+	/**
+	 * Path to the fonts storage
+	 *
+	 * @var string
+	 */
+	private $path;
+
+	/**
+	 * Version of the fonts.
+	 *
+	 * @var int
+	 */
+	private $version;
+
+	/**
+	 * Instantiate the class
+	 *
+	 * @param string               $base_path Base path to the fonts storage.
+	 * @param WP_Filesystem_Direct $filesystem WP Filesystem instance.
+	 */
+	public function __construct( $base_path, $filesystem = null ) {
+		parent::__construct( is_null( $filesystem ) ? rocket_direct_filesystem() : $filesystem );
+
+		$this->path = $base_path . get_current_blog_id() . '/';
+	}
+
+	/**
+	 * Hash the url
+	 *
+	 * @param string $font_url Font url.
+	 *
+	 * @return string
+	 */
+	private function hash_url( string $font_url ): string {
+		return md5( $font_url );
+	}
+
+	/**
+	 * Write font css to path
+	 *
+	 * @param string $font_url The font url to save locally.
+	 * @param string $provider The url of the page.
+	 *
+	 * @return bool
+	 */
+	public function write_font_css( string $font_url, string $provider ): bool {
+		$font_provider_path = $this->get_font_provider_path( $provider );
+		$hash_url           = $this->hash_url( $font_url );
+		$file               = $this->get_fonts_full_path( $font_provider_path, $hash_url );
+		$css_file_name      = $file . '.css';
+		$relative_path      = $this->get_fonts_relative_path( $font_provider_path, $hash_url );
+
+		if ( ! rocket_mkdir_p( dirname( $file ) ) ) {
+			return false;
+		}
+
+		$start_time = microtime( true );
+
+		$css_content = $this->download_font( html_entity_decode( $font_url ) );
+
+		if ( ! $css_content ) {
+			return false;
+		}
+
+		preg_match_all( '/url\((https:\/\/[^)]+)\)/', $css_content, $matches );
+		$font_urls = $matches[1];
+		$local_css = $css_content;
+
+		foreach ( $font_urls as $font_url ) {
+			$parsed_url = wp_parse_url( $font_url );
+			$font_path  = $parsed_url['path'];
+			$local_path = $file . $font_path;
+			$local_dir  = dirname( $local_path );
+
+			if ( ! rocket_mkdir_p( $local_dir ) ) {
+				continue;
+			}
+
+			if ( ! file_exists( $local_path ) ) {
+				$font_content = $this->download_font( $font_url );
+
+				if ( ! $font_content ) {
+					Logger::debug( 'Font download was not successful', [ 'Host Fonts Locally' ] );
+					continue;
+				}
+
+				$this->write_file( $local_path, $font_content );
+			}
+
+			$local_url = content_url( $relative_path . $font_path );
+			$local_css = str_replace( $font_url, $local_url, $local_css );
+		}
+
+		$end_time = microtime( true );
+		$duration = $end_time - $start_time;
+
+		// Add for test purpose.
+		Logger::debug( "Font download and optimization duration in seconds -- $duration", [ 'Host Fonts Locally' ] );
+
+		return $this->write_file( $css_file_name, $local_css );
+	}
+
+	/**
+	 * Download font from external url
+	 *
+	 * @param string $url Url of the file to download.
+	 *
+	 * @return mixed
+	 */
+	private function download_font( string $url ) {
+		$content = wp_remote_retrieve_body(
```

Comment:
This method is missing some checks, notably with the response code. You don't want to store a 404 response body into the font file.